### DISCUSSION TITLE
[MSIL] How to replace existing method by MSIL at runtime in CoreClr?

### DISCUSSION THREAD

**Discussion by Li7ye** [dotnet/runtime#110934] [Q&A] [answered]
Hello all,

[Harmony](https://harmony.pardeike.net/) package can support this feature, but I no idea to how implement. Ask for your help to tell me which knowledge I need to study, thanks.

**Comment by huoyaoyuan** [comment, marked as answer]
Modifying running code is not supported. There is hot patching that can update running code, but it's debug only and has many limitations.

Harmony itself doesn't do the patching. It only composes code blocks in to methods, and depends on [MonoMod](https://monomod.dev/docs/README.Core.html) to replace methods.

[MonoMod's documentation](https://monomod.dev/docs/Core/CoreCLRJitHooks.html) contains some discussion about hooking methods. It requires to update process memory directly. It's generally unsupported, unreliable and can be broken by many factors, including security or anti-malware enhancement, CLR feature update or refactor.

**Reply by Li7ye** [reply]
Noted and thanks. Do you have any good reference (e.g. book, documentation) to study CoreCLR and MSIL?

**Reply by huoyaoyuan** [reply]
Coreclr is still evolving. The best study material should be documentations and issues in this repo itself, to learn about existing and new features.
For MSIL there's not much thing to study. It's a simple stack machine language. Inspecting compiler outputs at https://sharplab.io is often sufficient.

**Comment by AustinWise** [comment]
One supported way of modifying MSIL at runtime is the [Profiler API](https://learn.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/profiling-overview). The docs say:

> The profiling API enables a profiler to change the in-memory CIL code stream for a routine before it is JIT-compiled.

The profiling API, as the name suggests, is designed for implementing tools like profilers or code coverage tools. It is a bit of an advanced API, so personally I would try to avoid rewriting IL at runtime if possible. Other types of code generation and rewriting are a easier to implement, including:

* Generating C# before build type.
* Using [source generators](https://learn.microsoft.com/dotnet/csharp/roslyn-sdk/) to generate C# at build time.
* Using IL rewriters (things like ILMerge) to rewrite IL after build time.
